### PR TITLE
Add a recursive merging strategy

### LINKED ISSUE #53: Allow deep merging of records when building a tree
Currently the merging of records only works for top level keys.  For example this:

```perl
  {
      families => [ {
          husband => 'Fred',
          wife    => 'Wimla',
      }, ],
      year => 1960,
  }
```

Merged with this:

```perl
    {
        families => [ {
            wife    => 'Wilma',
            child   => 'Pebbles',
        }, {
            husband => 'Barney',
            wife    => 'Betty',
            child   => 'Bamm-Bamm',
        }, ],  
        company => 'Hanna-Barbera Productions',
    }
```

Gives this:

```perl
    {
        families => [ {
            wife    => 'Wilma',
            child   => 'Pebbles',
        }, {
            husband => 'Barney',
            wife    => 'Betty',
            child   => 'Bamm-Bamm',
        }, ],
        year => 1960,  
        company => 'Hanna-Barbera Productions',
    }
```

Not

```perl
    {
        families => [ {
            husband => 'Fred',
            wife    => 'Wilma',    # note replaced value
            child   => 'Pebbles',
        }, {
            husband => 'Barney',
            wife    => 'Betty',
            child   => 'Bamm-Bamm',
        }, ],
        year => 1960,
        company => 'Hanna-Barbera Productions',
    }
```

This should probably be an optional feature

### PR DESCRIPTION
**Pull request by oschwald**
Closes #53.

### CODE REVIEW

File: lib/MaxMind/DB/Writer/Tree.pm [hunk 1]
```diff
@@ -263,13 +305,18 @@ sub write_tree {
...
-        filename              => { isa => 'Str' },
-        map_key_type_callback => { isa => 'CodeRef' },
-        database_type         => { isa => 'Str', optional => 1 },
-        description           => { isa => 'HashRef[Str]', optional => 1 },
+        filename                => { isa => 'Str' },
+        map_key_type_callback   => { isa => 'CodeRef' },
+        database_type           => { isa => 'Str', optional => 1 },
+        description             => { isa => 'HashRef[Str]', optional => 1 },
+        merge_record_collisions => { isa => 'Bool', optional => 1 },
+        merge_strategy => { isa => $MergeStrategyEnum, optional => 1 }
```
Comment (2shortplanks) [inline lib/MaxMind/DB/Writer/Tree.pm]: This doesn't look tidy

Comment (oschwald) [inline lib/MaxMind/DB/Writer/Tree.pm]: `tidyall` disagrees. I think it is like that as it doesn't want to wrap the hashref.

File: c/ppport.h [hunk 2]
```diff
@@ -1171,7 +1171,7 @@ av_reify|||
 av_shift|||
 av_store|||
 av_tindex||5.017009|
-av_top_index||5.017009|
+av_len||5.017009|
```
Comment (2shortplanks) [inline c/ppport.h]: Was this a manual change or the result of updating this file with the latest version from CPAN?

Comment (oschwald) [inline c/ppport.h]: This may have been accidental. Let me revert the changes to this file.

File: c/tree.c [hunk 3]
```diff
@@ -589,36 +592,113 @@ SV *merge_hashes_for_keys(MMDBW_tree_s *tree, const char *const key_from,
...
+
+    // Note that av_len() is really the index of the last element. In newer
+    // Perl versions, it is also called av_top_index() or av_tindex()
+    SSize_t from_top_index = av_len(from_array);
+    SSize_t into_top_index = av_len(into_array);
+
+    SSize_t new_top_index = from_top_index >
+                            into_top_index ? from_top_index : into_top_index;
+
+    AV *new_array = newAV();
```
Comment (2shortplanks) [inline c/tree.c]: Maybe we should be pre-allocating the size of the AV we're creating here with av_extend and then use av_store later rather than av_push?  This would avoid re-allocating arrays as stuff is pushed into them / wasting space with bigger arrays than needed

File: lib/MaxMind/DB/Writer/Tree.pm [hunk 4]
```diff
@@ -287,12 +331,12 @@ sub new_from_frozen_tree {
...
     $params->{database_type} = $database_type if defined $database_type;
     $params->{description}   = $description   if defined $description;
+    $params->{merge_strategy} = $merge_strategy if defined $merge_strategy;
 
     my $tree = _thaw_tree(
         $filename,
         $params_size + 4,
-        map { $params->{$_} }
-            qw( ip_version record_size merge_record_collisions ),
+        map { $params->{$_} } qw( ip_version record_size merge_strategy ),
```
Comment (2shortplanks) [inline lib/MaxMind/DB/Writer/Tree.pm]: So, this code introduces a subtle backward compatibility for stuff frozen with the old version of the code and unfrozen with the new in the situation where merge_record_collision was set to `1` in the frozen tree and you set merge_strategy to `none`.

Do we need to worry about this given frozen trees aren't considered compatible?  We should either address the problem or put a note in the release notes

File: lib/MaxMind/DB/Writer/Tree.pm [hunk 5]
```diff
@@ -429,10 +473,87 @@ By default, when an insert collides with a previous insert, the new data
...
+        year => 1960,
+        company => 'Hanna-Barbera Productions',
+    }
+
+=back
+
+In all merge strategies attempting to merge two differing data structures
+causes an exception.
+
+This parameter is optional. It defaults to C<toplevel>.
```
Comment (2shortplanks) [inline lib/MaxMind/DB/Writer/Tree.pm]: Doesn't this default to `none`?

File: c/tree.c (hunk 3 above)
Comment (oschwald) [inline c/tree.c]: Not sure it will make much of a difference in this case. It seems like a micro-optimization.

File: lib/MaxMind/DB/Writer/Tree.pm (hunk 4 above)
Comment (oschwald) [inline lib/MaxMind/DB/Writer/Tree.pm]: I am pretty sure we are the only users of the frozen trees. I guess we could check if `merge_record_collisions` exists and, if so, set `merge_strategy` to `toplevel` if it is not explicitly set  though.

File: lib/MaxMind/DB/Writer/Tree.pm (hunk 5 above)
Comment (oschwald) [inline lib/MaxMind/DB/Writer/Tree.pm]: Yeah, I think I just forgot to update this when I modified your docs.

Comment (oschwald) [inline lib/MaxMind/DB/Writer/Tree.pm]: Fixed in 69efeb6.

File: lib/MaxMind/DB/Writer/Tree.pm (hunk 4 above)
Comment (oschwald) [inline lib/MaxMind/DB/Writer/Tree.pm]: Fixed in 69efeb6.
